doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+    | Slist0sep (sym, sep) -> sprintf "LIST0 %sSEP %s" (prod_to_str ~plist [sym]) (prod_to_str ~plist [sep])
+    | Sopt sym -> sprintf "OPT %s" (prod_to_str ~plist [sym])
+    | Sparen sym_list -> sprintf "( %s) " (prod_to_str sym_list)
+    | Sprod sym_list -> sprintf "[ %s] "
+        (String.concat ""
+          (List.mapi (fun i r -> sprintf "%s%s" (if (i > 0) then "| " else "") (prod_to_str r)) sym_list))
+    | Sedit s -> sprintf "%s " s
+    (* todo: make PLUGIN output conditional on the set of prods? *)
+    | Sedit2 ("PLUGIN", s2) ->
+      if plist then
+        sprintf "     (%s plugin) " s2
+      else
+        sprintf "     (* %s plugin *) " s2
+    | Sedit2 (s, s2) -> sprintf "%s \"%s\" " s s2
+
+and prod_to_str_r plist prod =
+  match prod with
+  | p :: tl ->
+    let need_semi =
+      match prod with
+      | Snterm "IDENT" :: Sterm _ :: _
+      | Snterm "IDENT" :: Sprod _ :: _ -> true
+      | _ -> false in
+    (output_prod plist need_semi p) :: (prod_to_str_r plist tl)
+  | [] -> []
+
+and prod_to_str ?(plist=false) prod =
+  String.concat "" (prod_to_str_r plist prod)
+
+(* print a production *)
+let print_prod out plist nt prod first =
+  let pfx =
+    if first then
+      sprintf "%s%s" nt (if plist then " :" else ": [\n|")
+    else
+      if plist then sprintf "%s :" (String.make (String.length nt) ' ') else "|"
+  in
+  fprintf out "%s %s\n" pfx (prod_to_str ~plist prod)
+
+
+let rec output_prodn = function
+  | Sterm s -> sprintf "%s " s
+  | Snterm s -> sprintf "@%s " s
+  | Slist1 sym -> sprintf "{+ %s} " (output_prodn sym)
+  | Slist1sep (sym, sep) -> sprintf "{+%s%s} " (output_prodn sep) (output_prodn sym)
+  | Slist0 sym -> sprintf "{* %s} " (output_prodn sym)
+  | Slist0sep (sym, sep) -> sprintf "{*%s%s} " (output_prodn sep) (output_prodn sym)
+  | Sopt sym -> sprintf "{? %s} " (output_prodn sym)
+  | Sparen sym_list -> sprintf "%s " (prod_to_prodn sym_list)
+  | Sprod sym_list -> sprintf "%s "
+      (String.concat ""
+        (List.mapi (fun i r -> sprintf "%s%s" (if (i > 0) then "| " else "") (prod_to_prodn r)) sym_list))
+  | Sedit s -> sprintf "%s " s
+  | Sedit2 ("PLUGIN", s2) -> ""
+  | Sedit2 (s, s2) -> sprintf "%s \"%s\" " s s2
+
+and prod_to_prodn prod = map_and_concat output_prodn prod
+
+
+let print_prod_prodn out nt prod first =
+  let pfx = if first then sprintf "\n%s:\n" nt else "" in
+  fprintf out "%s %s\n" pfx (prod_to_prodn prod)
+
+
+type fmt = [`MLG | `PRODLIST | `PRODN ]
+
+(* print a subset of the grammar with nts in the specified order *)
+let print_in_order out g fmt nt_order hide =
+  let pfunc, plist = match fmt with
+  | `MLG -> print_prod out false, false
+  | `PRODLIST -> print_prod out true, true
+  | `PRODN -> print_prod_prodn out, true
+  in
+  List.iter
+    (fun nt ->
+      let prods = try NTMap.find nt g.map
+                  with Not_found -> []
+      in
+      List.iteri (fun i p ->
+          if not (StringSet.mem nt hide) then pfunc nt p (i=0))
```

Comment:
Extrude this check, and don't discriminate on the index as explained in the comment above.